### PR TITLE
Delayed emails

### PR DESCRIPTION
**Pull request by jeabakker**
This allows a user to receive one notification with an overview of what
happened in a configured interval

ToDo
- [x] tests
- [x] docs

### CODE REVIEW

File: engine/classes/Elgg/Email/DelayedEmailService.php [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	protected $invoker;
+	
+	/**
+	 * Create a new service
+	 *
+	 * @param DelayedEmailQueueTable $queue_table the queue database table
+	 * @param EmailService           $email       email service
+	 * @param ViewsService           $views       views service
+	 * @param Translator             $translator  translator service
+	 * @param Invoker                $invoker     ivoker serivce
```
Comment (jdalsem) [inline engine/classes/Elgg/Email/DelayedEmailService.php:58]: ivoker

File: engine/classes/Elgg/Email/DelayedQueue/DatabaseRecord.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+				return (int) $this->row->$name;
+			
+			case 'delivery_interval':
+				return $this->row->$name;
+			
+			case 'data':
+				return unserialize($this->row->data);
+		}
+		
+		return false;
```
Comment (jdalsem) [inline engine/classes/Elgg/Email/DelayedQueue/DatabaseRecord.php:57]: why not throw an exception?

File: engine/classes/Elgg/Email/DelayedQueue/DatabaseRecord.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		return (bool) _elgg_services()->delayedEmailQueueTable->deleteRow($this->id);
+	}
+	
+	/**
+	 * Get the notification for this record
+	 *
+	 * @return null|Notification
+	 */
+	public function getNotification(): ?Notification {
+		$data = $this->data;
```
Comment (jdalsem) [inline engine/classes/Elgg/Email/DelayedQueue/DatabaseRecord.php:75]: ``return $data instanceof Notification ? $data : null;``

File: engine/classes/Elgg/Email/DelayedQueue/EnqueueHandler.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Elgg\Email\DelayedQueue;
+
+use Elgg\Notifications\Notification;
+
+/**
+ * Handle the queue-ing of delayed email notifications
```
Comment (jdalsem) [inline engine/classes/Elgg/Email/DelayedQueue/EnqueueHandler.php:8]: *queueing

File: engine/classes/Elgg/Notifications/SendEmailHandler.php [hunk 5]
```diff
@@ -19,7 +19,7 @@ class SendEmailHandler {
 	public function __invoke(\Elgg\Hook $hook) {
 		if ($hook->getValue() === true) {
 			// assume someone else already sent the message
-			return;
+			return false;
```
Comment (jdalsem) [inline engine/classes/Elgg/Notifications/SendEmailHandler.php:22]: uhh why?

File: engine/schema/migrations/20210412110921_create_delayed_email_queue_table.php [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	 * PRIMARY KEY (`id`),
+	 * ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+	 */
+    public function change() {
+    	if ($this->hasTable('delayed_email_queue')) {
+    		return;
+    	}
+    	
+    	// table
+    	$table = $this->table('delayed_email_queue'	, [
```
Comment (jdalsem) [inline engine/schema/migrations/20210412110921_create_delayed_email_queue_table.php:24]: huh? whitespace

File: engine/schema/migrations/20210412110921_create_delayed_email_queue_table.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	 * `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
+	 * `recipient_guid` bigint(20) unsigned NOT NULL,
+	 * `delivery_interval` varchar(255) NOT NULL,
+	 * `data` mediumblob NOT NULL,
+	 * `timestamp` int(11) NOT NULL,
+	 * PRIMARY KEY (`id`),
+	 * ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+	 */
+    public function change() {
+    	if ($this->hasTable('delayed_email_queue')) {
```
Comment (jdalsem) [inline engine/schema/migrations/20210412110921_create_delayed_email_queue_table.php:19]: indentation? in the whole file

File: engine/tests/phpunit/integration/Elgg/Database/DelayedEmailQueueTableIntegrationTest.php [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+	}
+	
+	public function testqueueEmail() {
+		$notification = $this->getTestNotification();
+		$recipient = $notification->getRecipient();
+		
+		// insert
+		$this->assertTrue($this->table->queueEmail($recipient->guid, 'daily', $notification));
+		
+		$dt = $this->table->getCurrentTime('+10 seconds');
```
Comment (jdalsem) [inline engine/tests/phpunit/integration/Elgg/Database/DelayedEmailQueueTableIntegrationTest.php:61]: also a test to get if time -10sec?

File: engine/tests/phpunit/unit/Elgg/Database/DelayedEmailQueueTableUnitTest.php [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	 */
+	public function up() {
+		$this->table = _elgg_services()->delayedEmailQueueTable;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function down() {
+		$dt = $this->table->getCurrentTime('+10 seconds');
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/DelayedEmailQueueTableUnitTest.php:33]: add -10 seconds test

File: mod/notifications/views/default/notifications/subscriptions.js [hunk 10]
```diff
@@ -34,6 +34,24 @@ define(['jquery', 'elgg/Ajax'], function($, Ajax) {
 		
 		$record.find('.elgg-subscription-methods input').prop('disabled', false);
 		$record.find('.elgg-subscription-details-toggle').removeClass('elgg-state-active');
+	});
+	console.log('a');
```
Comment (jdalsem) [inline mod/notifications/views/default/notifications/subscriptions.js:38]: eh

File: mod/notifications/views/default/notifications/subscriptions.js [hunk 11]
```diff
@@ -34,6 +34,24 @@ define(['jquery', 'elgg/Ajax'], function($, Ajax) {
...
+	// prevent email and delayed email from being enabled at the same time
+	$(document).on('change', '.elgg-subscription-record .elgg-input-checkbox', function() {
+		console.log(1);
```
Comment (jdalsem) [inline mod/notifications/views/default/notifications/subscriptions.js:41]: aah

File: mod/notifications/views/default/notifications/settings.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
+define(['jquery'], function($) {
+	// prevent email and delayed email from being enabled at the same time
+	$(document).on('change', '.elgg-notifications-settings .elgg-input-checkbox', function() {
```
Comment (jdalsem) [inline mod/notifications/views/default/notifications/settings.js:3]: could add `:checked` to selector

File: mod/notifications/views/default/notifications/subscriptions.js [hunk 13]
```diff
@@ -34,6 +34,24 @@ define(['jquery', 'elgg/Ajax'], function($, Ajax) {
 		
 		$record.find('.elgg-subscription-methods input').prop('disabled', false);
 		$record.find('.elgg-subscription-details-toggle').removeClass('elgg-state-active');
+	});
+	console.log('a');
+	// prevent email and delayed email from being enabled at the same time
+	$(document).on('change', '.elgg-subscription-record .elgg-input-checkbox', function() {
```
Comment (jdalsem) [inline mod/notifications/views/default/notifications/subscriptions.js:40]: could add `:checked` to selector

File: views/default/email/delayed_email/html.php [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+			]);
+		}
+		
+		$lis[] = elgg_format_element('li', [], $text);
+	}
+	
+	$body .= elgg_format_element('ul', [], implode(PHP_EOL, $lis));
+	$body .= PHP_EOL . PHP_EOL;
+}
+
```
Comment (jdalsem) [inline views/default/email/delayed_email/html.php:103]: how to deal empty body? should we not return blank and tackle in cron handler?

File: views/default/email/delayed_email/plaintext.php [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+		if ($entity instanceof ElggEntity) {
+			$output .= ' ' . $object->getURL();
+		}
+		
+		$output .= PHP_EOL;
+	}
+	
+	$output .= PHP_EOL . PHP_EOL;
+}
+
```
Comment (jdalsem) [inline views/default/email/delayed_email/plaintext.php:98]: deal empty body

**Review by jdalsem — CHANGES_REQUESTED** [review]

**Review by jdalsem — APPROVED** [review]

File: engine/classes/Elgg/Email/DelayedQueue/DatabaseRecord.php (hunk 2 above)
Comment (jeabakker) [reply engine/classes/Elgg/Email/DelayedQueue/DatabaseRecord.php]: not the place

File: mod/notifications/views/default/notifications/subscriptions.js (hunk 11 above)
Comment (jeabakker) [reply mod/notifications/views/default/notifications/subscriptions.js]: oops

File: mod/notifications/views/default/notifications/subscriptions.js (hunk 10 above)
Comment (jeabakker) [reply mod/notifications/views/default/notifications/subscriptions.js]: oops

**Pull request merged**